wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift

```diff
@@ -0,0 +1,53 @@
+//
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import CoreCryptoSwift
+
+// MARK: - Protocols
+
+public protocol SafeCoreCryptoProtocol {
+    func perform<T>(_ block: (CoreCryptoProtocol) throws -> T) rethrows -> T
+    func unsafePerform<T>(_ block: (CoreCryptoProtocol) throws -> T) rethrows -> T
+}
+
+public struct SafeCoreCrypto: SafeCoreCryptoProtocol {
+
+    private let coreCrypto: CoreCryptoProtocol
+    private let safeContext: SafeFileContext
+
+    init(coreCrypto: CoreCryptoProtocol, coreCryptoConfiguration: CoreCryptoConfiguration) {
+        self.coreCrypto = coreCrypto
+        let directoryPathUrl = URL(fileURLWithPath: coreCryptoConfiguration.path).deletingLastPathComponent()
+        self.safeContext = SafeFileContext(fileURL: directoryPathUrl)
+    }
+
+    public func perform<T>(_ block: (CoreCryptoProtocol) throws -> T) rethrows -> T {
+        var result: T
+        safeContext.acquireDirectoryLock()
```

Comment:
Is this blocking?

Reply:
nope, so the issue was I was setting the lock on the db file and not the directory

Reply:
I'm a bit confused, not sure what the issue you're referring to is. I mean, does execution stop on the current thread until the lock is acquired?

Reply:
ah sorry that was another issue I had... I didn't really notice